src/server/multi_test.cc

```diff
@@ -782,4 +782,77 @@ TEST_F(MultiEvalTest, MultiSomeEval) {
   EXPECT_THAT(brpop_resp, ArgType(RespExpr::NIL_ARRAY));
 }
 
+TEST_F(MultiEvalTest, MultiPullsEarly) {
+  absl::FlagSaver fs;
+  absl::SetFlag(&FLAGS_multi_exec_squash, false);
+
+  for (unsigned i = 0; i < 20; ++i) {
+    string key = StrCat("x", i);
+    LOG(INFO) << key << ": shard " << Shard(key, shard_set->size());
+  }
+
+  Run({"mget", "x5", "x8", "x9", "x13", "x16", "x17"});
+  ASSERT_EQ(1, GetDebugInfo().shards_count);
+
+  auto fb1 = pp_->at(1)->LaunchFiber(Launch::post, [&] {
+    // Runs multi on shard0 1000 times.
+    for (unsigned j = 0; j < 1000; ++j) {
+      Run({"multi"});
+      Run({"incrby", "x13", "1"});
+      Run({"incrby", "x16", "1"});
+      Run({"incrby", "x17", "1"});
+      Run({"exec"});
+    }
+  });
+
+  auto fb2 = pp_->at(2)->LaunchFiber(Launch::dispatch, [&] {
+    // Runs multi on shard0 1000 times.
+    for (unsigned j = 0; j < 1000; ++j) {
+      Run({"multi"});
+      Run({"incrby", "x5", "1"});
+      Run({"incrby", "x8", "1"});
+      Run({"incrby", "x9", "1"});
+      Run({"exec"});
+    }
+  });
+
+  auto check_triple = [](const RespExpr::Vec& arr, unsigned start) {
+    if (arr[start].type != arr[start + 1].type || arr[start + 1].type != arr[start + 2].type) {
+      return false;
+    }
+
+    if (arr[0].type == RespExpr::STRING) {
+      string s0 = arr[start].GetString();
+      string s1 = arr[start + 1].GetString();
+      string s2 = arr[start + 2].GetString();
+      if (s0 != s1 || s1 != s2) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  bool success = pp_->at(0)->Await([&]() -> bool {
+    for (unsigned j = 0; j < 1000; ++j) {
+      auto resp = Run({"mget", "x5", "x8", "x9", "x13", "x16", "x17"});
```

Comment:
1. I'd access them separately to reduce waiting delays, you check them by triplets either way

Reply:
there is a reason I do it like this. I do not know whether mget will be scheduled after the first multi or the second one but if it is scheduled after one of them I test that the invariants hold for both of them.